File: fotoapparat/src/main/java/io/fotoapparat/preview/Frame.java
package io.fotoapparat.preview;

import java.util.Arrays;

import io.fotoapparat.parameter.Size;

/**
 * Frame of the preview stream.
 */
public class Frame {

    /**
     * Size of the frame in pixels (before rotation).
     */
    public final Size size;

    /**
     * Image in NV21 format.
     */
    public final byte[] image;

    /**
     * Clockwise rotation of the image in degrees relatively to user.
     */
    public final int rotation;

    public Frame(Size size, byte[] image, int rotation) {
        this.size = size;
        this.image = image;
        this.rotation = rotation;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Frame frame = (Frame) o;

        return rotation == frame.rotation
                && size.equals(frame.size)
                && Arrays.equals(image, frame.image);
    }

    @Override
    public int hashCode() {
        int result = size.hashCode();
        result = 31 * result + Arrays.hashCode(image);
        result = 31 * result + rotation;
        return result;
    }

    @Override
    public String toString() {
        return "Frame{" +
                "size=" + size +
                ", image= array(" + image.length + ")" +
                ", rotation=" + rotation +
                '}';
    }

}

File: fotoapparat/src/main/java/io/fotoapparat/preview/Frame.kt
package io.fotoapparat.preview

import io.fotoapparat.parameter.Resolution
import java.util.*

/**
 * Frame of the preview stream.
 */
data class Frame(
        /**
         * Resolution of the frame in pixels (before rotation).
         */
        val size: Resolution,
        /**
         * Image in NV21 format.
         */
        val image: ByteArray,
        /**
         * Clockwise rotation of the image in degrees relatively to user.
         */
        val rotation: Int
) {

    override fun toString(): String {
        return "Frame{" +
                "size=" + size +
                ", image= array(" + image.size + ")" +
                ", rotation=" + rotation +
                '}'
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Frame

        if (size != other.size) return false
        if (!Arrays.equals(image, other.image)) return false
        if (rotation != other.rotation) return false

        return true
    }

    override fun hashCode(): Int {
        var result = size.hashCode()
        result = 31 * result + Arrays.hashCode(image)
        result = 31 * result + rotation
        return result
    }

}
